node/impl/full/multisig.go

```diff
@@ -0,0 +1,282 @@
+
+	return api.AddressPair{
+		IDAddress:     execreturn.IDAddress,
+		RobustAddress: execreturn.RobustAddress,
+	}, nil
+}
+
+func (a *MsigAPI) MsigPropose(ctx context.Context, msig address.Address, to address.Address, amt types.BigInt, src address.Address, method uint64, params []byte) (api.MsigProposeOutput, error) {
+
+	if msig == address.Undef {
+		return api.UndefMsigProposalOutput, xerrors.Errorf("must provide a multisig address for proposal")
+	}
+
+	if to == address.Undef {
+		return api.UndefMsigProposalOutput, xerrors.Errorf("must provide a target address for proposal")
+	}
+
+	if !types.BigPos(amt) {
+		return api.UndefMsigProposalOutput, xerrors.Errorf("must provide a positive amount for proposed send")
+	}
+
+	if src == address.Undef {
+		defaddr, err := a.WalletAPI.WalletDefaultAddress(ctx)
+		if err != nil {
+			return api.UndefMsigProposalOutput, err
+		}
+		src = defaddr
+	}
+
+	enc, actErr := actors.SerializeParams(&samsig.ProposeParams{
+		To:     to,
+		Value:  amt,
+		Method: abi.MethodNum(method),
+		Params: params,
+	})
+	if actErr != nil {
+		return api.UndefMsigProposalOutput, actErr
+	}
+
+	msg := &types.Message{
+		To:       msig,
+		From:     src,
+		Value:    types.NewInt(0),
+		Method:   builtin.MethodsMultisig.Propose,
+		Params:   enc,
+		GasLimit: 100000,
+		GasPrice: types.NewInt(1),
+	}
+
+	smsg, err := a.MpoolAPI.MpoolPushMessage(ctx, msg)
+	if err != nil {
+		return api.UndefMsigProposalOutput, nil
+	}
+
+	fmt.Println("send proposal in message: ", smsg.Cid())
+
+	wait, err := a.StateAPI.StateWaitMsg(ctx, smsg.Cid())
+	if err != nil {
+		return api.UndefMsigProposalOutput, nil
+	}
+
+	if wait.Receipt.ExitCode != 0 {
+		return api.UndefMsigProposalOutput, fmt.Errorf("proposal returned exit %d", wait.Receipt.ExitCode)
+	}
+
+	_, v, err := cbg.CborReadHeader(bytes.NewReader(wait.Receipt.Return))
+	if err != nil {
+		return api.UndefMsigProposalOutput, nil
+	}
+
+	return api.MsigProposeOutput{
+		MessageCid:    smsg.Cid(),
+		TransactionID: v,
+	}, nil
+}
+
+func (a *MsigAPI) MsigApprove(ctx context.Context, msig address.Address, txID uint64, proposer address.Address, to address.Address, amt types.BigInt, src address.Address, method uint64, params []byte) (cid.Cid, error) {
+	return a.MsigApproveOrCancel(ctx, api.MsigApprove, msig, txID, proposer, to, amt, src, method, params)
+}
+
+func (a *MsigAPI) MsigCancel(ctx context.Context, msig address.Address, txID uint64, proposer address.Address, to address.Address, amt types.BigInt, src address.Address, method uint64, params []byte) (cid.Cid, error) {
+	return a.MsigApproveOrCancel(ctx, api.MsigCancel, msig, txID, proposer, to, amt, src, method, params)
+}
+
+func (a *MsigAPI) MsigApproveOrCancel(ctx context.Context, operation api.MsigProposeResponse, msig address.Address, txID uint64, proposer address.Address, to address.Address, amt types.BigInt, src address.Address, method uint64, params []byte) (cid.Cid, error) {
```

Comment:
This should be private. Assuming I understand Go correctly, changing the name to `msigApproveOrCancel` should do the trick.

Reply:
Was not aware of private methods in Go lol... Good call.